utils/enum_option.hh

```diff
@@ -0,0 +1,106 @@
+
+/*
+ * Copyright (C) 2015 ScyllaDB
+ */
+
+/*
+ * This file is part of Scylla.
+ *
+ * Scylla is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Scylla is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Scylla.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+// TODO: upstream this to Boost.
+
+#pragma once
+
+#include <boost/program_options/errors.hpp>
+#include <iostream>
+#include <sstream>
+#include <type_traits>
+
+/// A Boost program option holding an enum value.
+///
+/// The options parser will parse enum values with the help of the Mapper class, which provides a mapping
+/// between some parsable form (eg, string) and the enum value.  For example, it may map the word "January" to
+/// the enum value JANUARY.
+///
+/// Mapper must have a static method `map()` that returns a map from a streamable key type (eg, string) to the
+/// enum in question.
+///
+/// Example:
+///
+/// struct Type {
+///   enum ty { a1, a2, b1 };
+///   const static unordered_map<string, ty>& map();
+/// };
+/// const unordered_map<string, Type::ty>& Type::map() {
+///   static const auto inst =
+///     new unordered_map<string, Type::ty>{{"a1", Type::a1}, {"a2", Type::a2}, {"b1", Type::b1}};
+///   return *inst;
+/// }
+/// int main(int ac, char* av[]) {
+///   namespace po = boost::program_options;
+///   po::options_description desc("Allowed options");
+///   desc.add_options()
+///     ("val", po::value<enum_option<Type>>(), "Single Type")
+///     ("vec", po::value<vector<enum_option<Type>>>()->multitoken(), "Type vector");
```

Comment:
>     1. Is it possible to make it a set instead of a vector, to avoid duplicates and to allow the code to easily find out which option was enabled?

No, because `set` here doesn't compile.  It violates some static assert deep in Boost.

>     2. What is multitoken()?

It's what allows multiple option values: https://theboostcpplibraries.com/boost.program_options

Reply:
I actually might remove even `vector` here, because it's an unnecessary complication.  I'll explain in the blurb that this holds a single value (just like `po::value<int>` holds just a single `int`, not a list of them).